service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java

```diff
@@ -0,0 +1,111 @@
+package bio.terra.user.app.controller;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+import bio.terra.common.iam.SamUser;
+import bio.terra.user.testutils.BaseUnitTest;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+
+class ProfileApiControllerTest extends BaseUnitTest {
+  private static final String API = "/api/profile";
+
+  @Autowired private MockMvc mockMvc;
+
+  @MockBean private ProfileSamUserFactory userFactory;
+
+  private final SamUser user = mock(SamUser.class);
+
+  @BeforeEach
+  void beforeEach() {
+    when(userFactory.from(any())).thenReturn(user);
+    var id = String.valueOf(Math.random()).substring(2);
```

Comment:
We usually have a helper method to generate random numbers, that way you don't have to type this every time

https://cs.github.com/DataBiosphere/terra-workspace-manager/blob/792780ab6d1bf38ee9e4b811811952135974c7f2/integration/src/main/java/scripts/testscripts/FolderLifecycle.java?q=appendrandomnumber#L74

Though in this case, does subject have to be different every time? Can you do

```
when(user.getSubjectId()).thenReturn("fake-12345");
```

Reply:
+1, no need for random.  normally would only do this on some kind of integration test when you need to avoid conflicts with other tests

Reply:
@melissachang Since it's using a real db I wasn't sure if there would be conflicts between tests. Should I put a cleanup step in the beforeEach instead?

Reply:
In general, it's good practice to make your test resilient. Other tests could be running at the same time, other tests could have or have not run before or after this.

For example for WSM local development, I'll reuse a workspace between runs (as opposed to creating a new workspace for each test, which is what normally happens.) That workspace will have leftover resources from previous runs.

You don't need to clean up db. Instead, I think it's simpler to do something like this:

```
  @Test
  void recursiveSet() throws Exception {
    String path = TestUtils.appendRandomNumber("user.name.first");
    setUserProfile(path, "John");
    assertUserProfile(path, "John");
  }
```

Reply:
I think in that case it would be best to separate the tests by using random users and thus entire profile objects rather than randomizing the paths, because if several tests use the same parent nodes like `user` then they could interfere even if the leaves like `.name.first` are randomized.